### PR TITLE
Expose a public API for building RQNames

### LINKED ISSUE #55: Expose RQNameService.TryBuildForPublicAPIs to enable language interop in VS.
Some Visual Studio features use RQNames to communicate symbols.  It would be nice if there was a way to get the RQName for an ISymbol publicly.

### PR DESCRIPTION
**Pull request by Pilchie**
* Rename RQNameService to RQNameBuilder and make public.
* Rename TryBuildForPublicApi to Build and have it return null.
* Remove TryBuild, since we *only* build RQNames for public Api's now.
* Clean up implementation to not pass flag around for buildForPublicApi
* Remove RQDynamicType
* Create a singleton for System.Object.
* Add a comment explaining why we erase dynamic.

Fixes #55.

### CODE REVIEW

**Comment by Pilchie** [comment]
@mattwar - are you happy with the API shape?
@dpoeschl - okay with the simplification around `buildForPublicAPI`

**Comment by mattwar** [comment]
The type should be called RQName, and the method should be called Create or From.
It should not appear in a namespace called Implementation.
Is there not also an API that converts back to a symbol?

**Comment by Pilchie** [comment]
Thanks!  Renamed `Build` to `From` and moved to the root `Microsoft.VisualStudio.LanguageServices` namespaces.

There is no API that converts back, because we don't ever take RQNames as input - we just give them out when we are doing rename/go to definition.

**Comment by Pilchie** [comment]
Not saying we couldn't/shouldn't add a method to go back to `ISymbol`, just that we don't have one right now.

**Comment by mattwar** [comment]
:+1:

File: src/VisualStudio/Core/Def/Implementation/RQName/Nodes/RQType.cs [hunk 1]
```diff
@@ -1,11 +1,16 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using Microsoft.VisualStudio.LanguageServices.Implementation.RQName.SimpleTree;
+using System;
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/RQName/Nodes/RQType.cs]: Wrong sorting?

File: src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs [hunk 2]
```diff
@@ -1,25 +1,23 @@
...
 
 using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.Implementation.RQName;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.RQName
+namespace Microsoft.VisualStudio.LanguageServices
 {
-    internal static class RQNameService
+    /// <summary>
+    /// Helpers related to <see cref="RQName"/>s.
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs]: Should we explain what RQNames are here (and cref to IVsRefactorNotify)? It'd be nice if the Metadata view of IVsRefactorNotify included the remarks from the [msdn docs](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.ivsrefactornotify.aspx) with the grammar and examples.

File: src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs [hunk 3]
```diff
@@ -1,25 +1,23 @@
...
+    /// </summary>
+    public static class RQName
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs]: Update the file name to match

File: src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs [hunk 4]
```diff
@@ -1,25 +1,23 @@
...
+        /// <param name="symbol">The symbol to build an RQName for.</param>
+        public static string From(ISymbol symbol)
         {
             var node = RQNodeBuilder.Build(symbol);
-            rqname = (node != null) ? ParenthesesTreeWriter.ToParenthesesFormat(node.ToSimpleTree()) : null;
-
-            return node != null;
-        }
-
-        public static bool TryBuildForPublicAPIs(ISymbol symbol, out string rqname)
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs]: Very happy to see this go. :smile:

**Comment by dpoeschl** [comment]
:+1:

File: src/VisualStudio/Core/Def/Implementation/RQName/Nodes/RQType.cs (hunk 1 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/RQName/Nodes/RQType.cs]: Yep, fixed

File: src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs (hunk 2 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs]: Added the crefs.

File: src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs (hunk 3 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/RQName/RQNameService.cs]: Done.

**Pull request merged**